### ISSUE TITLE
Persistent Disk mount tests start failing a few hours after a cluster is created

### ISSUE THREAD

**Issue by quinton-hoole** [GoogleCloudPlatform/kubernetes#7972]
After running the standard e2e tests repeatedly against a long-living standard e2e test cluster, the following tests started to fail consistently.  Before starting to fail, they succeeded 4 times.  I'm busy gathering the cluster logs and debugging.  I assume that one or both of the nodes in the test cluster got into a state where they were unable to schedule pods, or perhaps mount the required persistent disk.

```
Identified problems

PD should schedule a pod w/ a RW PD, remove it, then schedule it on another host

/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/pd.go:122
Expected error:
    <*errors.errorString | 0xc2090e8de0>: {
        s: "gave up waiting for pod pd-test-767f4220-f599-11e4-929a-42010af01555 to be running after 300.00 seconds",
    }
    gave up waiting for pod pd-test-767f4220-f599-11e4-929a-42010af01555 to be running after 300.00 seconds
not to have occurred


PD should schedule a pod w/ a readonly PD on two hosts, then remove both.

/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/pd.go:187
Expected error:
    <*errors.errorString | 0xc209066720>: {
        s: "gave up waiting for pod pd-test-3dcc937e-f59a-11e4-929a-42010af01555 to be running after 300.00 seconds",
    }
    gave up waiting for pod pd-test-3dcc937e-f59a-11e4-929a-42010af01555 to be running after 300.00 seconds
not to have occurred
```

**Comment by quinton-hoole** [comment]
On the bright side, it's worth noting that these are the only two tests that have failed against the long-living cluster thus far, after 24 runs of 36 e2e tests each (i.e. >800 tests).  So this failure is very specific to persistent disks.  In other respects the long-living "soak" test cluster still seems perfectly healthy.

**Comment by quinton-hoole** [comment]
Aah yes.  Found this in kubelet.log:

I0508 05:40:54.709496    2993 event.go:203] Event(api.ObjectReference{Kind:"Pod", Namespace:"default", Name:"pd-test-b972ba0f-f544-11e4-9f96-42010af01555", UID:"c56bbef3-f544-11e4-ac4f-42010af0727d", APIVersion:"v1beta3", ResourceVersion:"2789", FieldPath:""}): reason: 'failedMount' Unable to mount volumes for pod "pd-test-b972ba0f-f544-11e4-9f96-42010af01555_default": Could not attach disk: Timeout after 10s
E0508 05:40:54.733744    2993 pod_workers.go:102] Error syncing pod c56bbef3-f544-11e4-ac4f-42010af0727d, skipping: Could not attach disk: Timeout after 10s
I0508 05:40:54.734401    2993 event.go:203] Event(api.ObjectReference{Kind:"Pod", Namespace:"default", Name:"pd-test-b972ba0f-f544-11e4-9f96-42010af01555", UID:"c56bbef3-f544-11e4-ac4f-42010af0727d", APIVersion:"v1beta3", ResourceVersion:"2789", FieldPath:""}): reason: 'failedSync' Error syncing pod, skipping: Could not attach disk: Timeout after 10s
I0508 05:40:57.714562    2993 manager.go:1242] Need to restart pod infra container for "pd-test-b972ba0f-f544-11e4-9f96-42010af01555_default" because it is not found
I0508 05:40:57.717831    2993 event.go:203] Event(api.ObjectReference{Kind:"Pod", Namespace:"default", Name:"pd-test-b972ba0f-f544-11e4-9f96-42010af01555", UID:"c56bbef3-f544-11e4-ac4f-42010af0727d", APIVersion:"v1beta3", ResourceV

**Comment by quinton-hoole** [comment]
Here's where the failure is happening:

https://github.com/GoogleCloudPlatform/kubernetes/blob/4a6ab7fd3698c5cc7172c5bd63efab320d1f2d17/pkg/volume/gce_pd/gce_util.go#L61

**Comment by quinton-hoole** [comment]
In summary we try to attach the GCE PD here:

https://github.com/GoogleCloudPlatform/kubernetes/blob/06c2f4e3d50d35e50ecb4a7909bec62d5d366875/pkg/cloudprovider/gce/gce.go#L636

If that fails, we do a somewhat optimistic check here to see whether the PD is already attached to this node/instance (according to GCE's "Get Node"):

https://github.com/GoogleCloudPlatform/kubernetes/blob/06c2f4e3d50d35e50ecb4a7909bec62d5d366875/pkg/cloudprovider/gce/gce.go#L645

It's worth noting that if the attachment of the PD fails for any reason in this process, e.g.

https://github.com/GoogleCloudPlatform/kubernetes/blob/06c2f4e3d50d35e50ecb4a7909bec62d5d366875/pkg/cloudprovider/gce/gce.go#L640

... it is never retried, as far as I can tell.

https://github.com/GoogleCloudPlatform/kubernetes/blob/b52f8656342c59f583b403faba8e5d9c61dcd4f2/pkg/volume/gce_pd/gce_pd.go#L200

Also note this juicy TODO over here:

// TODO: handle failed mounts here.
https://github.com/GoogleCloudPlatform/kubernetes/blob/b52f8656342c59f583b403faba8e5d9c61dcd4f2/pkg/volume/gce_pd/gce_pd.go#L189

I have other stuff to do right now, so I'm going to leave this one for V1.0 triage on Monday.

In summary, to fix this, we need to restructure the code to handle attach and mount failures, as there appear to be absolutely no retries in the code.

**Comment by quinton-hoole** [comment]
It's worth noting that the above does not completely explain away why all subsequent attempts to attach and mount a PD on this node are failing. I would suggest addressing the requirement for retries mentioned above, and at the same time improving the logging so that we can see exactly what is failing and how.

Superficially it looks like all of the functions in the call stack to attach and mount the PD to the node and pod are reporting success, but subsequent attempts to stat the mounted PD fail here:
https://github.com/GoogleCloudPlatform/kubernetes/blob/4a6ab7fd3698c5cc7172c5bd63efab320d1f2d17/pkg/volume/gce_pd/gce_util.go#L61

Two suggested strategies here:

1. Double-check that no possible errors are being incorrectly suppressed in the call stack.
2. If we fail to stat the supposedly attached and mounted PD, then do a full unmount, detach cycle, and try again to attach and mount it (a few times).

**Comment by quinton-hoole** [comment]
cc: @thockin who wrote the original code by the looks of things, in case he has any suggestions.

**Comment by thockin** [comment]
I am guilty of no such crime :)  I just moved it around, so my fingerprints
are on it.

And I have no insight to offer except to debug harder.

**Comment by quinton-hoole** [comment]
#6399 may well be related.

**Comment by swagiaal** [comment]
@quinton-hoole if the disk is failing to attach it is possible that the issue is outside of Kubernetes and in GCE. Look in the logs for something like "Error syncing pod [...] skipping: googleapi: [...]"
if there is nothing like that and the disk is attached then perhaps kube is failing to see the device. Try listing the device directory on the test machine 'ls /dev/disk/by-id/". Given error you pointed out that is likely what is happening but I cannot think of why :/
BTW even though AttachAndMountDisk does not have a tight retry loop it is called again in the next sync loop until the disk is attached and mounted

**Comment by vishh** [comment]
@quinton-hoole: Are you still looking into this issue? (Node team triage). 
cc @saad-ali

**Comment by quinton-hoole** [comment]
I did a little bit of code archeology yesterday, but got interrupted before I got to the bottom of the problem.  I'll unassign myself for now, and update this bug if I find anything.  I definitely won't look at it again before tomorrow.

**Comment by saad-ali** [comment]
I'm looking in to this. I am able to repro on my own cluster by deploying a cluster, verifying that the PD test passes, leaving the cluster unused overnight, and then running PD test the following morning.

**Comment by saad-ali** [comment]
The test creates a PersistentDisk (PD) when it starts. It then creates a pod with the PD as a volume which causes Kubernetes to attach the PD. The Kubernetes code attaches the PD to the node that the pod is running apparently without issue. However, although the PD appears to attach properly (confirmed in dev console UI) it never shows up under `/dev/disk/by-id/`, so the test times out waiting for it. Following up internally with PD eng to hopefully figure out why.

```
PD 
  should schedule a pod w/ a RW PD, remove it, then schedule it on another host
...
STEP: creating PD
INFO: Creating new PD. pdName="e2e-c35e8b41-05a0-11e5-8b46-f0921cde18c1"
....
e2e-c35e8b41-05a0-11e5-8b46-f0921cde18c1  Standard persistent disk  10 GB us-central1-b e2e-test-saadali-minion-689a
[PD IS SUCCESSFULLY  CREATED]
...
[PD IS SUCCESSFULLY  ATTACHED TO e2e-test-saadali-minion-689a]:
e2e-test-saadali-minion-689a  us-central1-b e2e-test-saadali-minion-689a, e2e-c35e8b41-05a0-11e5-8b46-f0921cde18c1  
...
689a Logs filled with :
I0529 01:22:41.633896   32203 gce_util.go:74] os.Stat("/dev/disk/by-id/google-e2e-c35e8b41-05a0-11e5-8b46-f0921cde18c1") returned error: stat /dev/disk/by-id/google-e2e-c35e8b41-05a0-11e5-8b46-f0921cde18c1: no such file or directory
I0529 01:22:41.633913   32203 gce_util.go:74] os.Stat("/dev/disk/by-id/scsi-0Google_PersistentDisk_e2e-c35e8b41-05a0-11e5-8b46-f0921cde18c1") returned error: stat /dev/disk/by-id/scsi-0Google_PersistentDisk_e2e-c35e8b41-05a0-11e5-8b46-f0921cde18c1: no such file or directory
```

**Comment by thockin** [comment]
Thanks Saad.

**Comment by saad-ali** [comment]
Tracking internally with b/21597665

**Comment by quinton-hoole** [comment]
@saad-ali Are you still working on this?

**Comment by saad-ali** [comment]
Yep, I'll follow up with you offline.

**Comment by quinton-hoole** [comment]
b/21597665  (Google-internal reference)

**Comment by pmorie** [comment]
FYI @markturansky

**Comment by thockin** [comment]
Saad, can you update this if you get a second?

**Comment by saad-ali** [comment]
Sure. I implemented the suggested work around in PR #9929 but we decided that was too broadscope (hitting all the devices on the system with a `udevadm trigger`). So I closed that and created follow-up PR #10169 which tries to be much finer scoped and also implements retries on detach which should also help with issue #9994.

**Comment by quinton-hoole** [comment]
These tests are still failing in our continuous soak tests.  Re-opening...

e.g. jenkins:  /job/kubernetes-soak-continuous-e2e/1845/

**Comment by saad-ali** [comment]
Looks like these failures are caused by: `Pod cannot be started due to exceeded capacity`

Which is not the same problem as the bugs identified and fixed in this issue or issue #9994 (namely disks not getting attached or detached).

So let's track this in a separate issue: I opened #10899
